### PR TITLE
Sortable: added checks for the helper touching the edge of the containment while dragging

### PR DESCRIPTION
**Pull request by lukeapage**
Rebased from https://github.com/jquery/jquery-ui/pull/1060 - I'll handle any comments to get this in as @mickylad is busy at the moment

Fixed #5772 - Sortable: containment parent restricting replacement of first and last elements

This issue can be reproduced based on where the mouse is positioned when sorting - in the patched versions you can drag from anywhere in the item to trigger a rearrangement, where as in the unpatched versions you must drag such that the initial mouse down position inside the item will overlap with the target item.

There is also an existing issue where the vertical container will be pushed down when sorting into the last index, which is not fixed in this pull request.

### CODE REVIEW

File: tests/unit/sortable/sortable_options.js [hunk 1]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
...
+
+	equal(item.index(), 0, "Item is sorted to first position");
+
+	item.simulate("drag", { dx: 150, dy: 150 });
+
+	equal(item.index(), 8, "Item is sorted to last position");
 });
 
-test("{ cursor: 'auto' }, default", function() {
+/*test("{ cursor: 'auto' }, default", function() {
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: typo

**Comment by In4matik** [comment]
Thanks for PR! 

> There is also an existing issue where the vertical container will be pushed down when sorting into the last index, which is not fixed in this pull request.
Any update or PR about this one?

**Comment by lukeapage** [comment]
@In4matik No.. We've been waiting a very long time to have this merged (since 16th Aug 2013!!)

There isn't much incentive to further fix issues until this is merged or even commented on.

**Comment by mikesherov** [comment]
@lukeapage sorry for the long delay, and thanks for contributing! There are a lot of style guide violations here, which I can clean up myself while landing this, but you might want to take a look at : http://contribute.jquery.org/style-guide/js/ if you're going to contribute more patches.

Also, please sign our CLA: http://contribute.jquery.org/CLA so I can land this patch!

Thanks again for contributing!

**Comment by lukeapage** [comment]
Hi,

@morchard did the work and has signed the CLA.
I rebased and have also signed the CLA (I have previously contributed some documentation and test fix PR's which have been merged)

Please can you point out the style violations? Do you not use jscs for style checking (and that has passed)? We've had a look and can't see anything obvious, but it would be good to learn what is a problem.

File: tests/unit/sortable/sortable_options.js [hunk 2]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
 
 test("{ containment: Selector }", function() {
 	ok(false, "missing test - untested code is broken code.");
+});*/
+
+test("#5772: wide element intersect sorting to first and last position on x-axis", function() {
+	expect(2);
+
+	var element = $("#sortable-horizontal").sortable({
+				axis: "x",
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: Is is that this should have only 1 level of indentation?

File: ui/sortable.js [hunk 3]
```diff
@@ -374,47 +375,67 @@ return $.widget("ui.sortable", $.ui.mouse, {
...
+				}
 
-				if (this.options.tolerance === "pointer" || this._intersectsWithSides(item)) {
-					this._rearrange(event, item);
-				} else {
+				// cannot intersect with itself
+				// no useless actions that have been done before
+				// no action if the item moved is the parent of the item checked
+				if (itemElement !== this.currentItem[0] &&
+					this.placeholder[intersection === 1 ? "next" : "prev"]()[0] !== itemElement &&
```
Comment (lukeapage) [inline ui/sortable.js]: this should have an extra 1 layer of indentation?

File: tests/unit/sortable/sortable_options.js [hunk 4]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
...
+	item.simulate("drag", { dx: 150, dy: 150 });
+
+	equal(item.index(), 8, "Item is sorted to last position");
+});
+
+test("#5772: element pointer sorting to first and last position on grid", function() {
+	expect(2);
+
+	var element = $("#sortable-grid").sortable({
+				containment: "parent",
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: this should be only 1 indent?

File: tests/unit/sortable/sortable_options.js [hunk 5]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
...
+	item.simulate("drag", { dy: 150 });
+
+	equal(item.index(), 4, "Item is sorted to last position");
+});
+
+test("#5772: element intersect sorting to first and last position on grid", function() {
+	expect(2);
+
+	var element = $("#sortable-grid").sortable({
+				containment: "parent",
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: this should only be 1 indent?

File: tests/unit/sortable/sortable_options.js [hunk 6]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
...
+	item.simulate("drag", { dy: 150 });
+
+	equal(item.index(), 4, "Item is sorted to last position");
+});
+
+test("#5772: tall element pointer sorting to first and last position on y-axis", function() {
+	expect(2);
+
+	var element = $("#sortable-vertical").sortable({
+				axis: "y",
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: this should only be 1 indent?

File: tests/unit/sortable/sortable_options.js [hunk 7]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
...
+	item.simulate("drag", { dx: 150 });
+
+	equal(item.index(), 4, "Item is sorted to last position");
+});
+
+test("#5772: tall element intersect sorting to first and last position on y-axis", function() {
+	expect(2);
+
+	var element = $("#sortable-vertical").sortable({
+				axis: "y",
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: this should only be 1 indent?

File: tests/unit/sortable/sortable_options.js [hunk 8]
```diff
@@ -212,9 +212,127 @@ test("{ containment: 'window' }", function() {
...
+	item.simulate("drag", { dx: 150 });
+
+	equal(item.index(), 4, "Item is sorted to last position");
+});
+
+test("#5772: wide element pointer sorting to first and last position on x-axis", function() {
+	expect(2);
+
+	var element = $("#sortable-horizontal").sortable({
+				axis: "x",
```
Comment (lukeapage) [inline tests/unit/sortable/sortable_options.js]: this should only be 1 indent?

File: ui/sortable.js [hunk 9]
```diff
@@ -374,47 +375,67 @@ return $.widget("ui.sortable", $.ui.mouse, {
...
-				this.placeholder[intersection === 1 ? "next" : "prev"]()[0] !== itemElement &&
-				!$.contains(this.placeholder[0], itemElement) &&
-				(this.options.type === "semi-dynamic" ? !$.contains(this.element[0], itemElement) : true)
-			) {
+		if(touchingContainmentEdge !== false &&
+				this.placeholder[0] !== this.items[touchingContainmentEdge].item[0]) {
+			this._rearrange(event, this.items[touchingContainmentEdge]);
+			this._trigger("change", event, this._uiHash());
+		} else {
+			//Rearrange
```
Comment (lukeapage) [inline ui/sortable.js]: needs line above

**Comment by mikesherov** [comment]
Yeah, mostly around spacing, actually. Unfortunately, we've yet to do a bulk JSCS cleanup yet and so haven't turned it on for the messier code in the codebase.

As a maintainer of JSCS myself, it pains me to say that.

**Comment by lukeapage** [comment]
@mikesherov any progress? Are there any outstanding style violations to fix? I've fixed the ones I found.

**Comment by mikesherov** [comment]
@lukeapage, we're releasing 1.12 soon. This patch will definitely be in there.

**Comment by victor-homyakov** [comment]
`_mouseDrag` already has 120 lines of code and three levels of nesting (`for() { if() { if() {...} } }`).

Adding ~20 more lines and one more nesting level makes this even more harder to read and understand (not to mention further fixes/improvements).

Maybe it's time to split it into smaller parts/extract some helper functions?

**Comment by lukeapage** [comment]
@victor-homyakov this patch was first proposed August 2013 and has been lingering around since then with no reason why it hasn't been pulled.

It's probably a fair point, but I think it shouldn't be part of this PR.

**Comment by cornem** [comment]
We would really like to see this PR included in one of the library's next releases. Why has this fix been hanging around for so long? Could someone please provide an update on this, or a possible work-around?

**Comment by scottgonzalez** [comment]
Unfortunately @mikesherov never got around to merging this for 1.12. I'm going to close this due to the conflicts, but if you're still interested in getting the patches merged, please file a new PR.

Sorry about the lack of feedback from the team.

**Pull request closed without merging**